File: C1.UWP.FlexChart/VB/DrillDown/Layer/TreemapDataLayer.vb
Public Class TreemapDataLayer
    Inherits Bindable
    Private _itemsSource As Leaf()
    Private _depth As Integer = 0

    Private _histories As New ObservableCollection(Of KeyValuePair(Of String, Leaf()))()

    Public Sub New(dataSource As Leaf())
        ItemsSource = dataSource
        AddHandler _histories.CollectionChanged, AddressOf _histories_CollectionChanged
        _histories.Add(New KeyValuePair(Of String, Leaf())("All Types", dataSource))

        Dim actionBack As Action(Of String) = Sub(s As String)
                                                  Back(s)
                                              End Sub
        BackCommand = New RelayCommand(actionBack)
    End Sub

    Private Sub _histories_CollectionChanged(sender As Object, e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)
        Notify("CurrentType")
    End Sub

    Public Property ItemsSource() As Leaf()
        Get
            Return _itemsSource
        End Get
        Set
            SetProperty(_itemsSource, Value, "ItemsSource")
        End Set
    End Property

    Public Property Depth() As Integer
        Get
            Return _depth
        End Get
        Set
            SetProperty(_depth, Value, "Depth")
        End Set
    End Property

    Public Property Histories() As ObservableCollection(Of KeyValuePair(Of String, Leaf()))
        Get
            Return _histories
        End Get
        Set
            SetProperty(_histories, Value, "Histories")
        End Set
    End Property

    Public ReadOnly Property CurrentType() As String
        Get
            Return Histories.Last().Key
        End Get
    End Property

    Public Property BackCommand() As RelayCommand
        Get
            Return m_BackCommand
        End Get
        Set
            m_BackCommand = Value
        End Set
    End Property
    Private m_BackCommand As RelayCommand

    Public Overridable Sub Back(obj As Object)
        Dim idx = Integer.Parse(obj.ToString())
        For i As Integer = _histories.Count() - 1 To idx + 1 Step -1
            Histories.RemoveAt(i)
        Next
        Depth = idx
        ItemsSource = _histories.Last().Value
    End Sub

    Public Sub DrillDown(item As Object)
        Dim dataItem As Leaf = TryCast(item, Leaf)
        If dataItem IsNot Nothing AndAlso dataItem.Items IsNot Nothing AndAlso dataItem.Items.Count() > 0 Then
            Histories.Add(New KeyValuePair(Of String, Leaf())(dataItem.Type, dataItem.Items))
            ItemsSource = dataItem.Items
            Depth += 1
        End If
    End Sub
End Class

File: C1.UWP.FlexChart/CS/DrillDown/Layer/TreemapDataLayer.cs
using System.Linq;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace DrillDown
{
    public class TreemapDataLayer : Bindable
    {
        private Leaf[] _itemsSource;
        private int _depth = 0;
        
        private ObservableCollection<KeyValuePair<string, Leaf[]>> _histories = new ObservableCollection<KeyValuePair<string, Leaf[]>>();

        public TreemapDataLayer(Leaf[] dataSource)
        {
            ItemsSource = dataSource;
            _histories.CollectionChanged += _histories_CollectionChanged;
            _histories.Add(new KeyValuePair<string, Leaf[]> ( "All Types", dataSource ));
            BackCommand = new RelayCommand(p => Back(p));
        }

        private void _histories_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            Notify("CurrentType");
        }

        public Leaf[] ItemsSource
        {
            get { return _itemsSource; }
            set
            {
                SetProperty(ref _itemsSource, value, "ItemsSource");
            }
        }

        public int Depth
        {
            get { return _depth; }
            set
            {
                SetProperty(ref _depth, value, "Depth");
            }
        }

        public ObservableCollection<KeyValuePair<string, Leaf[]>> Histories
        {
            get { return _histories; }
            set
            {
                SetProperty(ref _histories, value, "Histories");
            }
        }

        public string CurrentType
        {
            get { return Histories.Last().Key; }
        }

        public RelayCommand BackCommand
        {
            get; set;
        }

        public virtual void Back(object obj)
        {
            var idx = int.Parse(obj.ToString());
            for (int i = _histories.Count() - 1; i>idx; i--)
            {
                Histories.RemoveAt(i);
            }
            Depth = idx;
            ItemsSource = _histories.Last().Value;
        }

        public void DrillDown(object item)
        {
            var dataItem = item as Leaf;
            if(dataItem != null && dataItem.Items != null && dataItem.Items.Count() > 0)
            {
                Histories.Add(new KeyValuePair<string, Leaf[]>(dataItem.Type, dataItem.Items));
                ItemsSource = dataItem.Items;
                Depth++;
            }
        }
    }
}
